src/khoj/database/adapters/__init__.py

```diff
@@ -1031,6 +1037,52 @@ async def aget_advanced_conversation_config(user: KhojUser = None):
             return server_chat_settings.chat_advanced
         return await ConversationAdapters.aget_default_conversation_config(user)
 
+    @staticmethod
+    async def aget_server_webscraper():
+        server_chat_settings = await ServerChatSettings.objects.filter().prefetch_related("web_scraper").afirst()
+        if server_chat_settings is not None and server_chat_settings.web_scraper is not None:
+            return server_chat_settings.web_scraper
+        return None
+
+    @staticmethod
+    async def aget_enabled_webscrapers():
+        enabled_scrapers = []
+        server_webscraper = await ConversationAdapters.aget_server_webscraper()
+        if server_webscraper:
+            # Only use the webscraper set in the server chat settings
+            enabled_scrapers = [
+                (server_webscraper.type, server_webscraper.api_key, server_webscraper.api_url, server_webscraper.name)
+            ]
+        if not enabled_scrapers:
+            # Use the enabled web scrapers, using the newest created scraper first, until get web page content
+            enabled_scrapers = [
+                (scraper.type, scraper.api_key, scraper.api_url, scraper.name)
+                async for scraper in WebScraper.objects.all().order_by("-created_at").aiterator()
+            ]
+        if not enabled_scrapers:
+            # Use scrapers enabled via environment variables
+            if os.getenv("FIRECRAWL_API_KEY"):
+                api_url = os.getenv("FIRECRAWL_API_URL", "https://api.firecrawl.dev")
+                enabled_scrapers.append(
+                    (WebScraper.WebScraperType.FIRECRAWL, os.getenv("FIRECRAWL_API_KEY"), api_url, "Firecrawl")
+                )
+            if os.getenv("OLOSTEP_API_KEY"):
+                api_url = os.getenv("OLOSTEP_API_URL", "https://agent.olostep.com/olostep-p2p-incomingAPI")
+                enabled_scrapers.append(
+                    (WebScraper.WebScraperType.OLOSTEP, os.getenv("OLOSTEP_API_KEY"), api_url, "Olostep")
+                )
+
+            # Jina is the default fallback scrapers to use as it does not require an API key
+            api_url = os.getenv("JINA_READER_API_URL", "https://r.jina.ai/")
+            enabled_scrapers.append((WebScraper.WebScraperType.JINA, os.getenv("JINA_API_KEY"), api_url, "Jina"))
+
+            # Only enable the direct web page scraper by default in self-hosted single user setups.
+            # Useful for reading webpages on your intranet.
+            if state.anonymous_mode or in_debug_mode():
+                enabled_scrapers.append((WebScraper.WebScraperType.DIRECT, None, None, "Direct"))
+
+        return enabled_scrapers
```

Comment:
Can we create a cache for this result? Unless the server setting is changed, it should always return the same `enabled_scrapers`. We can have a post `save` signal on the `WebScraper` to trigger `aget_enabled_webscrapers`, though this won't work in a distributed (multi-container) environment. We can also use time-based invalidation. What do you think? I'm okay with also merging this in as-is only because the frequency of scraping shouldn't be terribly high and it's just making two calls to the DB.

Reply:
Good point (on caching this and using save to invalidate cache). But maybe better as a future optimization?